### DISCUSSION TITLE
multicast on W10 not working

### DISCUSSION THREAD

**Discussion by gartnl** [dannagle/PacketSender#237] [Q&A]
I'm trying to send a multicast from W10 to 224.0.0.251 port 5353. But it fails. Wether I join or do not join a multicast-group does not matter.
I can send a broadcast and unicast. W10 is sending multicasts when requesting a .local site. 
I did disable the firewall but that did not make a difference ( and the outgoing firewall isn't enabled anyway on a standard install ).

The listening device is an ESP32, it just displays any arriving udp packet. All devices are displayed ( including the aggressive chromecast ones ). No packets are blocked between wired lan and WiFi.

PS version = 7.2.3

So what am I missing? Any suggestions are welcome.

Henk

**Comment by dannagle** [comment]
Just a couple of tips...

- You do not have to join multicast to send to a multicast address. 
If Packet Sender is able to send plain unicast to your device, it should be able to send multicast (if your device is configured to listen correctly). Just tell Packet Sender to send a plain UDP data packet to 224.0.0.251 port 5353

- Joining multicast is not enough. You need to bind port a port also to get the actual data. 
Typically, the same socket used for join also binds.

- Wireshark may help your troubleshooting:  https://www.wireshark.org/
You set Wireshark to capture UDP

I do not have an ESP32 device. If you have access to 2 computers with 2 Packet Senders, I can offer assistance towards a proof-of-concept you could then try to migrate. I also own a Raspberry Pi if that is a decent equivalent for what you are trying to accomplish.

**Reply by gartnl** [reply]
Thanks for your response. I believe I've solved my problem.

First: the ESP was programmed to simply display all udp packets on port 5353 on a serial monitor ( my old laptop can't cope with wireshark ).

So I monitored all udp traffic. I did see udp from my main machine from browser requests or pinging from the command prompt. I did see multicast traffic from all other devices.
And I did see broadcasts from packetsender.  Multi/unicasts however did not appear.
Using yet another machine multicasts from that machine were just fine , packetsender did send them.
( and udp between the machines worked off course )

This led me to take a second look at my main machine: the problem was the virtual ethernet adapter used by hyper-v.
Binding packetsender to the correct ip for my local network showed all multicast on the esp-monitor.
Disabling the virtual ethernet adapter lets packetsender work without binding the local ip4 address.
You can even enable it again so it becomes second in the list of adapters, and packetsender is just fine using the first.

So, problem solved: disable the virtual adapter or bind your ip to packetsender. It also solved the problem I had that led me to use packetsender: the Arduino IDE is now able to send multicasts to OTA programmable devices.

What remains: the underlying reason for this behavior still eludes me. Should a multicast not be sent out to every available network?

regards

Henk

**Reply by dannagle** [reply]
No, multicast and UDP does not need to go out on every available network. A troubleshooting strategy I often recommend is disabling all network cards except the one you know you want. 

There are a couple ways you can bind a socket. 

- General bind (just the port number). This is the most common. Socket is now listening on a particular port on all available network addresses. 
- IP bind. The notation looks something like 10.22.54.25:553. This is useful if you want to isolate a service to a single IP on a single network adapter. This is also how you can avoid port conflicts that often arise from a general bind. The main reason for this is well-known ports (like Apache). You can have multiple instances of Apache isolated to different IP addresses all wanting port 80 and 443. 

That's all good, but back to your question: how do you send? If you use general bind, the network the UDP goes out is semi-random. I say "semi" because it will probably consistently choose the same one. However, which initial one it chooses is unknown. 

To make it non-random, you need to do an IP-bind. If you bind your socket to 10.22.54.25:553, when you send a packet, the packet will always go out whatever network card has the IP address 10.22.54.25. 

If you want to send out to ALL network cards, you need to make a socket for all your available IP addresses and bind each one and then send. It is tedious, but that is the way to do it reliably. 

Hope that helps. Glad your problem got fixed.

**Reply by gartnl** [reply]
"No, multicast and UDP does not need to go out on every available network."

And that's were I went wrong. Another lesson learned. Thanks for your info and time!
